### PR TITLE
Implement OAuth data deletion on Tenant Deletion

### LINKED ISSUE #1388: Implement deletion of OAuth data on Tenant Deletion
**Description:**
All the OAuth Consumer Apps created under a tenant should get deleted in the tenant deletion process.

**Related Issues:**
wso2/carbon-multitenancy#146

### PR DESCRIPTION
**Pull request by sumedhe**
### Proposed changes in this pull request
Deletion of OAuth consumer apps and OIDC properties, in the tenant deletion process.

Fixes #1388

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminServiceImpl.java [hunk 1]
```diff
@@ -865,6 +865,22 @@ public void removeOAuthApplicationData(String consumerKey) throws IdentityOAuthA
...
+     * @throws IdentityOAuthAdminException
+     */
+    public void removeAllOAuthApplicationData(int tenantId) throws IdentityOAuthAdminException {
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Deleting all OAuth Application data of the tenant: " + tenantId);
+        }
+
+        OAuthAppDAO dao = new OAuthAppDAO();
+        dao.removeConsumerApplicationsByTenantId(tenantId);
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminServiceImpl.java:881]: What about the applications cached?

Comment (mefarazath) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminServiceImpl.java:881]: Also, we need to check what happens to the OAuth app -> Service Provider associations with this delete.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java [hunk 2]
```diff
@@ -695,6 +695,32 @@ public void removeConsumerApplication(String consumerKey) throws IdentityOAuthAd
...
+    /**
+     * Delete all consumer applications of a given tenant.
+     *
+     * @param tenantId Id of the tenant
+     * @throws IdentityOAuthAdminException
+     */
+    public void removeConsumerApplicationsByTenantId(int tenantId) throws IdentityOAuthAdminException {
+
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection(false)) {
+            try (PreparedStatement prepStmt = connection
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java:707]: these 2 try blocks can be combined.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOAuthTenantMgtListener.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import org.wso2.carbon.identity.oauth.IdentityOAuthAdminException;
+import org.wso2.carbon.identity.oauth2.internal.OAuth2ServiceComponentHolder;
+import org.wso2.carbon.stratos.common.beans.TenantInfoBean;
+import org.wso2.carbon.stratos.common.exception.StratosException;
+import org.wso2.carbon.stratos.common.listeners.TenantMgtListener;
+
+/**
+ * Identity OAuth Tenant Management Listener.
+ */
+public class IdentityOAuthTenantMgtListener implements TenantMgtListener {
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/listener/IdentityOAuthTenantMgtListener.java:30]: Can't we reuse the TenantCreationEventListener?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminServiceImpl.java (hunk 1 above)
Comment (IsuraD) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminServiceImpl.java:881]: Do we need to separately delete OAuth apps?  OAuth apps should be auto-deleted when we delete the Service Providers. So, This step is not required. 

@mefarazath WDYT?

Comment (IsuraD) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminServiceImpl.java:881]: I think, there we only delete from SP tables. Thus let's keep this.  As Farasath mentioned, we have to delete the associations too.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminServiceImpl.java [hunk 4]
```diff
@@ -865,6 +865,22 @@ public void removeOAuthApplicationData(String consumerKey) throws IdentityOAuthA
 
     }
 
+    /**
+     * Remove all OAuth consumer applications of a tenant.
+     *
+     * @param tenantId Id of the tenant
```
Comment (Kanapriya) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminServiceImpl.java:871]: Change this as follows:
@param tenantId of the tenant

Comment (sumedhe) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminServiceImpl.java:871]: I have placed it there as, `tenantId` is the parameter name and *Id of the tenant* is the description. Hope that's fine.

**Pull request merged**